### PR TITLE
Fix ts errors

### PR DESCRIPTION
**Pull request by Qard**
Fixes #2693
Fixes #2787

### What does this PR do?

Takes #2693 and fixes remaining TypeScript errors. I've also included the minor docs change in #2787.

### Motivation

There's lots of TypeScript users. We should better support them.

### Security 
Datadog employees:
- [ ] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -83,7 +83,8 @@ export declare interface Tracer extends opentracing.Tracer {
    * unless there is already an active span or `childOf` option. Note that this
    * option is deprecated and has been removed in version 4.0.
    */
-  trace<T> (name: string, fn: (span?: Span, fn?: (error?: Error) => any) => T): T;
+  trace<T> (name: string, fn: (span: Span) => T): T;
+  trace<T> (name: string, fn: (span: Span, done: (error?: Error) => string) => T): T;
```
Comment (nsavoire) [inline index.d.ts:87]: It was mentioned in https://github.com/DataDog/dd-trace-js/pull/2693 that return type of `done` callback should be `void`.

Comment (Qard) [reply index.d.ts]: Yep, missed that. Should be fixed now. 😅

Comment (nsavoire) [reply index.d.ts]: Looks like test.ts [needs](https://github.com/DataDog/dd-trace-js/actions/runs/7385905479/job/20091536262?pr=3912) to be updated too 😄

Comment (Qard) [reply index.d.ts:87]: Ah, right...that means it's a breaking change so I just removed the change to void. We'll have to make a not to switch that return value in the next major. 🤔

Comment (Qard) [reply index.d.ts]: Actually, looking at it more carefully, other changes are technically major too so I'll just mark the whole PR as major and try to land that in v5.

File: docs/test.ts [hunk 2]
```diff
@@ -371,7 +372,7 @@ tracer.trace('test', { service: 'foo', resource: 'bar', type: 'baz' }, () => {})
 tracer.trace('test', { measured: true }, () => {})
 tracer.trace('test', (span: Span) => {})
 tracer.trace('test', (span: Span, fn: () => void) => {})
-tracer.trace('test', (span: Span, fn: (err: Error) => string) => {})
+tracer.trace('test', (span: Span, fn: (err: Error) => void) => {})
```
Comment (simon-id) [inline docs/test.ts:375]: that means the callback is supposed to return void right ?

Comment (Qard) [reply docs/test.ts:375]: Not an _explicit_ `return void`, if that's what you're asking. Just the absence of a return. It being a string before was _probably_ an error before but I don't know the history. Either way, it's been no-return for as long as I'm aware of, so the types should match.

Comment (simon-id) [reply docs/test.ts:375]: What about making it `any` ? sometimes people write stuff like:
```javascript
if (err) {
  return something(err)
}
```
and something() might return a non null

Comment (Qard) [reply docs/test.ts:375]: Sure, I made it `any`.

**Review by tlhunter — DISMISSED** [review]

File: docs/test.ts (hunk 2 above)
Comment (simon-id) [reply docs/test.ts:375]: I still see `void` here, did you push your commit ?

**Review by simon-id — DISMISSED** [review]

File: docs/test.ts (hunk 2 above)
Comment (Qard) [reply docs/test.ts:375]: I realized it would actually _always_ be `void`. The `done` function is coming from _us_ and it never has a return. 😅

**Review by simon-id — APPROVED** [review]

**Review by szegedi — APPROVED** [review]

**Pull request merged**